### PR TITLE
[MASIC] Make test_bgp_bounce.py support for multi-asic platforms

### PR DESCRIPTION
**Pull request by wenyiz2021**
### Description of PR

for test introduced from https://github.com/Azure/sonic-mgmt/pull/2866, test_bgp_bounce.py calls 'docker cp' cmd from host to bgp(s), via bgp_helpers.py, which does not consider multi-asic scenario, where we want to copy bgp config files to all BGPs. 
This PR does the following change:
1. For multi-asic, copy to all BGPs, as well as restart all BGPs and make sure all BGPs come back. For single-asic remains the same behavior.
2. Add restart_bgp() in multi_asic.py to do consideration for multi-asic/single-asic all in this file, and leave the wait_until() in bgp_helpers.py to let know if all BGPs are back up.
3. Original restart_bgp(asic_index) moved to restart_service_on_asic in multi_asic.py to work on one asic.

Summary:
Fixes # (issue)

### Type of change

- [x] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [x] 202012

#### What is the motivation for this PR?
Support test_bgp_bounce.py on multi-asic platform and some refactor.

#### How did you do it?
under sonic-mgmt/tests, do:
./run_tests.sh-n {multi_asic testbed} -i ../ansible/xxx,../ansible/veos  -f ../ansible/testbed.csv -u -e "--skip_sanity --disable_loganalyzer" -c bgp/test_bgp_bounce.py
#### How did you verify/test it?
Before:
2022-05-13T02:18:37.5367742Z E               "stderr": "Error: No such container:path: bgp:/usr/share/sonic/templates/bgpd/bgpd.conf.j2", 
2022-05-13T02:18:37.5370441Z E               "stderr_lines": [
2022-05-13T02:18:37.5373090Z E                   "Error: No such container:path: bgp:/usr/share/sonic/templates/bgpd/bgpd.conf.j2"
2022-05-13T02:18:37.5375680Z E               ], 

After:
bgp/test_bgp_bounce.py::test_bgp_bounce PASSED                                                                                                                                                                                                     [ 10%]
metadata-scripts/test_metadata_upgrade_path.py::test_cancelled_upgrade_path[warm-svcstr-n-acs-1] SKIPPED                                                                                                                                       [ 20%]
metadata-scripts/test_metadata_upgrade_path.py::test_upgrade_path[warm-svcstr-n-acs-1] SKIPPED                                                                                                                                                 [ 30%]
metadata-scripts/test_metadata_upgrade_path.py::test_warm_upgrade_sad_path[warm-sad-svcstr-n-acs-1] SKIPPED                                                                                                                                    [ 40%]
metadata-scripts/test_metadata_upgrade_path.py::test_warm_upgrade_sad_path[warm-multi_sad-svcstr-n-acs-1] SKIPPED                                                                                                                              [ 50%]
metadata-scripts/test_metadata_upgrade_path.py::test_warm_upgrade_sad_path[warm-sad_bgp-svcstr-n-acs-1] SKIPPED                                                                                                                                [ 60%]
metadata-scripts/test_metadata_upgrade_path.py::test_warm_upgrade_sad_path[warm-sad_lag_member-svcstr-n-acs-1] SKIPPED                                                                                                                         [ 70%]
metadata-scripts/test_metadata_upgrade_path.py::test_warm_upgrade_sad_path[warm-sad_lag-svcstr-n-acs-1] SKIPPED                                                                                                                                [ 80%]
metadata-scripts/test_metadata_upgrade_path.py::test_warm_upgrade_sad_path[warm-sad_vlan_port-svcstr-n=-acs-1] SKIPPED                                                                                                                          [ 90%]
metadata-scripts/test_metadata_upgrade_path.py::test_warm_upgrade_sad_path[warm-sad_inboot-svcstr-n-acs-1] SKIPPED                                                                                                                             [100%]

--------------------------------------------------------------------------------------------- generated xml file: /var/src/sonic-mgmt-int/tests/logs/tr.xml ----------------------------------------------------------------------------------------------
================================================================================================================ short test summary info =================================================================================================================
SKIPPED [9] /var/src/sonic-mgmt-int/tests/metadata-scripts/test_metadata_upgrade_path.py:39: base_image_list or target_image_list is empty
========================================================================================================= 1 passed, 9 skipped in 369.93 seconds ==========================================================================================================

### CODE REVIEW

File: tests/common/devices/multi_asic.py [hunk 1]
```diff
@@ -594,3 +601,8 @@ def docker_copy_from_asic(self, container_name, src, dst, asic_id = 0):
         if duthost.is_multi_asic:
             container_name += str(asic_id)
         self.shell("sudo docker cp {}:{} {}".format(container_name, src, dst))
+        
+    def restart_bgp(self):
```
Comment (arlakshm) [inline tests/common/devices/multi_asic.py:605]: why do we need to do `reset_service` every time ? Normally we need reset services only if services is not running properly

File: tests/bgp/bgp_helpers.py [hunk 2]
```diff
@@ -30,22 +30,27 @@ def apply_bgp_config(duthost, template_name):
...
+    duthost.restart_bgp()
+    # Check if all BGPs are up
+    if duthost.is_multi_asic:
+        for asic_index in range(duthost.facts["num_asic"]):
+            docker_name = duthost.asic_instance(asic_index).get_docker_name("bgp")
+            pytest_assert(wait_until(100, 10, 0, duthost.is_service_fully_started, docker_name), "BGP not started.")
+    else:
+        pytest_assert(wait_until(100, 10, 0, duthost.is_service_fully_started, "bgp"), "BGP not started.") 
 
 def restart_bgp(duthost, asic_index=DEFAULT_ASIC_ID):
```
Comment (arlakshm) [inline tests/bgp/bgp_helpers.py:43]: in this file in some places we use `restart_bgp` and in some places we use `duthost.restart_bgp`. Can we use one function everywhere

Comment (wenyiz2021) [reply tests/bgp/bgp_helpers.py:43]: I have thought of this, is it ok to check is_fully_started inside multi_asic.py or need to make it in bgp_helpers.py each time?

File: tests/common/devices/multi_asic.py (hunk 1 above)
Comment (wenyiz2021) [reply tests/common/devices/multi_asic.py:605]: This is refactored from original code, should I remove it?

File: tests/bgp/bgp_helpers.py (hunk 2 above)
Comment (wenyiz2021) [reply tests/bgp/bgp_helpers.py]: PR updated, can you please review gain?

File: tests/common/devices/multi_asic.py (hunk 1 above)
Comment (wenyiz2021) [reply tests/common/devices/multi_asic.py]: removed

File: tests/bgp/bgp_helpers.py [hunk 3]
```diff
@@ -30,22 +29,16 @@ def apply_bgp_config(duthost, template_name):
...
+    duthost.docker_copy_to_all_asics('bgp', template_name, DEFAULT_BGP_CONFIG)
+    duthost.restart_service("bgp")
+    # Check if all BGPs are up
+    if duthost.is_multi_asic:
+        for asic_index in range(duthost.facts["num_asic"]):
+            docker_name = duthost.asic_instance(asic_index).get_docker_name("bgp")
+            pytest_assert(wait_until(100, 10, 0, duthost.is_service_fully_started, docker_name), "BGP not started.")
+    else:
+        pytest_assert(wait_until(100, 10, 0, duthost.is_service_fully_started, "bgp"), "BGP not started.") 
+
```
Comment (arlakshm) [inline tests/bgp/bgp_helpers.py:41]: Can we add this block as function in multi_asic.py ?

Comment (wenyiz2021) [reply tests/bgp/bgp_helpers.py]: in multi_asic.py it is calling itself or the function in sonic_asic.py, to achieve that will need to add assert in the 2 restart_service() in the 2 files, but other functions do not have these checks, wonder if it is needed here?

    def restart_service(self, service):
        if service in self._DEFAULT_SERVICES:
            return self.sonichost.restart_service(service, service)
        for asic in self.asics:
            asic.restart_service(service)

**Review by arlakshm — APPROVED** [review]

**Pull request merged**